### PR TITLE
array_append function impl

### PR DESCRIPTION
**Pull request by disideris**
Resolves [#11907](https://github.com/crate/crate/issues/11907)

## Checklist

 - [ ] Added an entry in `CHANGES.txt` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by seut** [comment]
ok to test

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        assertNormalize("array_append([[1,2],[3,4]], [null, 6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), Arrays.asList(null,6))));
+    }
+
+    @Test
+    public void testEmptyArrayNullObject() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("The array argument of the `array_append` function can be of " +
+                                        "undefined type or of undefined inner type, but not when object argument " +
+                                        "is also undefined");
+        assertNormalize("array_append([], null)", null);
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:140]: What is the reason to not allow this?

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void testNullArrayNullObject() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("The array argument of the `array_append` function can be of " +
+                                        "undefined type or of undefined inner type, but not when object argument " +
+                                        "is also undefined");
+        assertNormalize("array_append(null, null)", null);
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:149]: Also here, why it shouldn't be allowed?

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void testNullArrayInnerTypeNullObject() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("The array argument of the `array_append` function can be of " +
+                                        "undefined type or of undefined inner type, but not when object argument " +
+                                        "is also undefined");
+        assertNormalize("array_append([null], null)", null);
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:158]: This also :-) What's the reason?

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        assertEvaluate("array_append([1])", null);
+    }
+
+    @Test
+    public void testThreeArguments() throws Exception {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Unknown function: array_append(_array(1), _array(2), _array(3))," +
+                                        " no overload found for matching argument types: (integer_array, integer_array, integer_array).");
+        assertEvaluate("array_append([1], [2], [3])", null);
+    }
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:183]: This is testing our general function matching logic and is already tested at the related components. I'd suggest to just remove these here.

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    private static final ArrayType<Integer> INTEGER_ARRAY_TYPE = new ArrayType<>(DataTypes.INTEGER);
+
+    @Test
+    public void testNormalizeWithValueSymbols() throws Exception {
+        assertNormalize("array_append([10, 20], 30)", isLiteral(Arrays.asList(10, 20, 30)));
+    }
+
+    @Test
+    public void testNormalizeWithTextValueSymbols() throws Exception {
+        assertNormalize("array_append(['10', '20'], '30')", isLiteral(Arrays.asList("10", "20", "30")));
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:46]: Any reason to add an explicit string test, isn't is using the same code path like any other array inner type?

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+    @Test
+    public void testCastTextToDoubleInnerTypes() throws Exception {
+        assertNormalize("array_append([1,2.0], '3.0')", isLiteral(Arrays.asList(1.0, 2.0, 3.0)));
+    }
+
+    @Test
+    public void testDifferentConvertableInnerTypesDouble() throws Exception {
+        assertNormalize("array_append([1::integer], 2::double)", isLiteral(Arrays.asList(1.0, 2.0)));
+    }
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:97]: I think one test to ensure that implicit casting of inner types is allowed for this scalar is enough.

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+    @Test
+    public void test2DArrayArgumentIncludingNullValueWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],[null, 4]], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(null,4), Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithArrayObjectIncludingNullValue() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], [null, 6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), Arrays.asList(null,6))));
+    }
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:132]: As the scalar implementation makes no difference if the array inner type is an array or not, I think these test case can be safely omitted.

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            Arrays.asList(1, null, 3, null),
+            Literal.of(Arrays.asList(1, null, 3), INTEGER_ARRAY_TYPE));
+    }
+
+    @Test
+    public void testTwoIntegerArguments() throws Exception {
+        assertEvaluate("array_append(int_array, 3)",
+            Arrays.asList(1, 2, 3),
+            Literal.of(Arrays.asList(1, 2), INTEGER_ARRAY_TYPE));
+    }
```
Comment (seut) [inline server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:204]: Based on existing scalar test cases it may be a bit confusing when and if at all to test with `asserNormalize` vs  `assertEvaluate`. Actually this is only needed if the scalar implements a concrete `normalizeSymbol` method, otherwise the `evaluate` method will be used for both, literals and references so there is no real reason to test both. I'd suggest to remote them.

File: docs/general/builtins/scalar-functions.rst [hunk 9]
```diff
@@ -2358,6 +2358,78 @@ Examples
...
+    +--------------------+
+    SELECT 1 row in set (... sec)
+
+.. NOTE::
+
+    Appending to arrays with array_append in updates is handy, but unfortunately
+    not isolated. We use optimistic concurrency control to ensure that your
+    update operation used the latest state of the row. But only 3 retry
+    attempts are made by fetching the newest version again and if they all
+    fail, the query fails.
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:2431]: I've just create a PR to improve the array documentation (specially related to this note and some examples of `array_cat`), see  https://github.com/crate/crate/pull/11932.
So I'd kindly ask you to adjust your new documentation in the same way or otherwise any feedback here (like reasons to keep it or add something else) is also more than welcome.

**Review by seut — COMMENTED** [review]
@disideris Thank you a lot for this contribution.

This looks really good, I've just added some suggestions related to test cases and also question about the `null` argument handling.

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 1 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: Appending a null value to an empty array indeed makes sense.
I tried to follow the pattern from array_cat when concatenating an empty array with another array with a null value.

select * from array_cat([], [null]);
SQLParseException[One of the arguments of the `array_cat` function can be of undefined inner type, but not both]

So it's better to allow array_append([], null)?

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 2 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:94]: Same as above

select * from array_cat(null, [null]);                                                                                                  
SQLParseException[One of the arguments of the `array_cat` function can be of undefined inner type, but not both]

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 3 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:103]: Same here

select * from array_cat([null], [null]);                                                                                                
SQLParseException[One of the arguments of the `array_cat` function can be of undefined inner type, but not both]

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 4 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java]: Removed :)

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 5 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java]: No, not really. Was just checking by hand string values and added some test. 
Removed :)

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 6 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java]: Done

File: docs/general/builtins/scalar-functions.rst (hunk 9 above)
Comment (disideris) [reply docs/general/builtins/scalar-functions.rst]: Added array_append doc to new scalar doc!

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 1 above)
Comment (seut) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: > So it's better to allow array_append([], null)?
Yes I think we should tweak `array_cat` instead.

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 2 above)
Comment (seut) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:94]: I see not technical reason to forbid this, so same here, we should tweak `array_cat` instead

File: server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java (hunk 1 above)
Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: So, how to handle this? 
1. Should I not use the ensure functions from ArrayArgumentValidators, make a new ensure function or check the validity inside the evaluate function of the implementation? 
2. Should i also change array_cat according to the choice of the 1 above?

Comment (seut) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: I'd suggest to just leave out the validator, everything else like ensuring both type are matching is already ensured by the function signature matching logic.
Right, for consistency we should change `array_cat` as well, but not sure if its worth doing now (null case maybe a bit of an edge case), I'd definitely would do this as part of this PR.

Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: I can make the changes by the end of the day.

Just for clarification:
I'll leave out the validator so that we can append a null value to an empty array, but not allowing to append anything to a null array?
array_append([], null) -> [null] valid
array_append(null, anyvalue) -> not valid or [anyvalue]?

For array_cat I understood that we shouldn't change it as part of this PR. 
So what do you mean by this?
> I'd definitely would do this as part of this PR

That's all I think to finalize the PR.

Comment (seut) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: Sorry mistake by me, I've meant "would **not** do this as part of this PR"

Comment (seut) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java:85]: > I'll leave out the validator so that we can append a null value to an empty array, but not allowing to append anything to a null array?

I'd leave out the validator at all and thus also allow to append anything to a null array.

Comment (disideris) [reply server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java]: ok! pushed it :)

File: server/src/main/java/io/crate/expression/scalar/ArrayAppendFunction.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public final List<Object> evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+        DataType<?> innerType = ((ArrayType<?>) this.boundSignature.getReturnType().createType()).innerType();
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/ArrayAppendFunction.java:74]: I think this could be moved into a instance variable to avoid resolving it on every row.

File: server/src/main/java/io/crate/expression/scalar/array/ArrayArgumentValidators.java [hunk 11]
```diff
@@ -58,4 +58,15 @@ public static void ensureSingleArgumentArrayInnerTypeIsNotUndefined(List<DataTyp
                 "When used with only one argument, the inner type of the array argument cannot be undefined");
         }
     }
+
+    public static void ensureArrayArgAndObjectArgAreNotBothUndefined(List<DataType<?>> dataTypes, String functionName) {
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/array/ArrayArgumentValidators.java:62]: Is this now unused? If so can we remove it?

**Review by seut — COMMENTED** [review]
I've added some final comments, otherwise this looks good.
May I also ask you to add a CHANGES entry? 
Thank you a lot.

File: server/src/main/java/io/crate/expression/scalar/ArrayAppendFunction.java (hunk 10 above)
Comment (disideris) [reply server/src/main/java/io/crate/expression/scalar/ArrayAppendFunction.java]: Done!

File: server/src/main/java/io/crate/expression/scalar/array/ArrayArgumentValidators.java (hunk 11 above)
Comment (disideris) [reply server/src/main/java/io/crate/expression/scalar/array/ArrayArgumentValidators.java]: yes sure, removed!

**Review by seut — COMMENTED** [review]
Looks fine now, thank you a lot for your effort.
Only the documentation changes require some modifications, see also the conflicts.
But we could also take over from here and fix these conflicts while merging, to not bother you more with nitty details :)

**Comment by BaurzhanSakhariev** [comment]
retest this please

**Review by seut — APPROVED** [review]
👍 @BaurzhanSakhariev thanks for taking over. please squash the commits before merging
